src/utils/read-rows-resumption.ts

```diff
@@ -0,0 +1,258 @@
+   * the chunk transformer, the set of row keys and row ranges is calculated and
+   * updated. The calculated row keys and ranges are used along with other
+   * properties provided by the user like limits and filters to compute and
+   * return a request that will be used in the next read rows call.
+   *
+   * @return {protos.google.bigtable.v2.IReadRowsRequest} The request options
+   * for the next readrows request.
+   */
+  getResumeRequest(): protos.google.bigtable.v2.IReadRowsRequest {
+    const reqOpts = this
+      .tableStrategyInfo as google.bigtable.v2.IReadRowsRequest;
+
+    // Create the new reqOpts
+    reqOpts.rows = {};
+
+    // TODO: preprocess all the keys and ranges to Bytes
+    reqOpts.rows.rowKeys = this.rowKeys.map(
+      Mutation.convertToBytes
+    ) as {} as Uint8Array[];
+
+    reqOpts.rows.rowRanges = this.ranges.map(range =>
+      Filter.createRange(
+        range.start as BoundData,
+        range.end as BoundData,
+        'Key'
+      )
+    );
+
+    if (this.options.filter) {
+      reqOpts.filter = Filter.parse(this.options.filter);
+    }
+
+    if (this.hasLimit) {
+      reqOpts.rowsLimit = this.rowsLimit - this.rowsRead;
+    }
+    return reqOpts;
+  }
+
+  /**
+   * Decides if the client is going to retry a request.
+   *
+   * canResume contains the logic that will decide if the client will retry with
+   * another request when it receives an error. This logic is passed along to
+   * google-gax and used by google-gax to decide if the client should retry
+   * a request when google-gax receives an error. If canResume returns true then
+   * the client will retry with another request computed by getResumeRequest. If
+   * canResume request returns false then the error will bubble up from gax to
+   * the handwritten layer.
+   *
+   * @param {GoogleError} [error] The error that Google Gax receives.
+   * @return {boolean} True if the client will retry
+   */
+  canResume(error: GoogleError): boolean {
+    // First update the row keys and the row ranges based on the last row key.
+    const lastRowKey = this.chunkTransformer
+      ? this.chunkTransformer.lastRowKey
+      : '';
+    if (lastRowKey) {
+      spliceRanges(this.ranges, lastRowKey);
+      this.rowKeys = getRowKeys(this.rowKeys, lastRowKey);
+    }
+    if (error.statusDetails === 'RetryInfo') {
+      return true;
+    }
+    // If all the row keys and ranges are read, end the stream
+    // and do not retry.
+    if (this.rowKeys.length === 0 && this.ranges.length === 0) {
+      return false;
+    }
+    // If there was a row limit in the original request and
+    // we've already read all the rows and met/exceeded that limit, end the
+    // stream and do not retry.
+    if (this.hasLimit && this.rowsLimit === this.rowsRead) {
+      return false;
+    }
+    if (
+      error.code &&
+      (RETRYABLE_STATUS_CODES.has(error.code) || isRstStreamError(error))
```

Comment:
Yes. See the line in `toRetryOptions` that says `gaxOpts?.retry?.retryCodes || [],`.

Reply:
Note the user passes in `gaxOpts`.

Reply:
How does that connect here though? I only see this comparing against the constant `RETRYABLE_STATUS_CODES`

Reply:
Because this function is not used if `gaxOpts?.retry?.shouldRetryFn` is provided. See the line of code that says `gaxOpts?.retry?.shouldRetryFn || canResume,` which creates `RetryOptions` that ignore `canResume` if `shouldRetryFn` is provided. Also, I made an update to have it ignore `canResume` if `retryCodes` are provided and use that instead so the user can specify `shouldRetryFn` or `retryCodes`. If you want to see an example of this working, check out the test case called `should use a different set of retry codes`.

Reply:
Actually, I thought about this again and we still want the check on empty row keys and empty ranges to be done if the user provides retry codes so I updated the code here to use retry codes if the user provides them instead of using the retry codes in RetryOptions and ignoring this function.

Reply:
Yes, that was what I was getting it. It seems concerning to me that this wasn't caught by any tests. Can you make sure the read-rows-resumption tests all pass with custom retryable errors?